src/Semantics/Interp.dfy

```diff
@@ -254,13 +285,69 @@ module Bootstrap.Semantics.Interp {
             case FunctionCall() =>
               InterpFunctionCall(e, env, argvs[0], argvs[1..])
           })
-      case Bind(vars, exprs: seq<Expr>, body: Expr) =>
-        var Return(vals, ctx) :- InterpExprs(exprs, env, ctx);
-        InterpBind(e, env, ctx, vars, vals, body)
+
+      case Bind(bvars, vals: seq<Expr>, body: Expr) =>
+        // This is ``InterpBind`` inlined (see the discussion for the ``Block`` case)
+        var vars := VarsToNames(bvars);
+        // A bind acts like a scope containing variable declarations:
+        // - Open a scope
+        var ctx1 := StartScope(ctx);
+        // - Evaluate the rhs
+        var Return(vals, ctx2) :- InterpExprs(vals, env, ctx1);
+        // - Save the shadowed variables to the rollback
+        var ctx3 := SaveToRollback(ctx2, vars);
+        // - Augment the context with the new bindings
+        var ctx4 := ctx3.(locals := AugmentContext(ctx3.locals, vars, vals));
+        // - Execute the body
+        var Return(v, ctx5) :- InterpExpr(body, env, ctx4);
+        // - End the scope
+        var ctx6 := EndScope(ctx, ctx5);
+        // Return
+        Success(Return(v, ctx6))
+
+      case VarDecl(vdecls, ovals) =>
+        var vars := VarsToNames(vdecls);
+        // Evaluate the rhs, if there is
+        if ovals.Some? then
+          var Return(vals, ctx) :- InterpExprs(ovals.value, env, ctx);
+          // Save the variables to the rollback context
+          var ctx := SaveToRollback(ctx, vars);
+          // Augment the context with the new bindings
+          var ctx := ctx.(locals := AugmentContext(ctx.locals, vars, vals));
+          // Continue
+          Success(Return(Unit, ctx))
+        else
+          // Save the variables to the rollback context
+          var ctx := SaveToRollback(ctx, vars);
+          // Continue
+          Success(Return(Unit, ctx))
+
+      case Update(vars, vals) =>
+        var Return(vals, ctx) :- InterpExprs(vals, env, ctx);
+        // DISCUSS: we need to check that the variables have been declared, but there is actually
+        // no way to do that for now (a variable declared but not initialized doesn't appear in the
+        // environment - a variable declaration may only update the rollback context). But is not
+        // checking that variables have been declared really a problem?
```

Comment:
I think it is a problem: accepting programs with undeclared variables means that we won't catch passes that fail to declare their variables.

Of course it's not an issue if we separately check that, but if we make this an invariant of the expr type then local transformations become complicated.

Reply:
Let's discuss that

Reply:
Let's table this for later; can you create an issue?